### DISCUSSION TITLE
How do I spawn an entity with a dynamic set of components?

### DISCUSSION THREAD

**Discussion by alice-i-cecile** [bevyengine/bevy#11409] [Q&A] [answered]
When working with Bevy, I often find myself wanting to spawn entities within a single function that may not have the same set of components.
However, doing so is more challenging than it should be!

Suppose we have:
```rust
use bevy::prelude::*;

#[derive(Component)]
struct A;

#[derive(Component)]
struct B;

#[allow(dead_code)]
enum ComponentStrategy {
    A,
    B,
    AAndB,
}
```
Then we want to do:
```rust
/// This is the obvious way to do it,
/// but simply doesn't work due to mismatched arm types
#[test]
fn impl_bundle_return_type() {
    fn spawn_bundle_naive(strategy: &ComponentStrategy) -> impl Bundle {
        match strategy {
            ComponentStrategy::A => (A,),
            ComponentStrategy::B => (B,),
            ComponentStrategy::AAndB => (A, B),
        }
    }
} 
```

How do I fix this?!

This question comes up regularly, and I have yet to see a good answer for it. So I made a [repo](https://github.com/alice-i-cecile/dynamic-components-experiments) exploring this problem. My answers are taken directly from there. Feel free to steal code from it directly: it's CC0. This has been discussed in issue format in #2157 and #3227: while I still think we should add direct support for this to Bevy, here are some solutions that work *today* (in Bevy 0.12).

**Comment by alice-i-cecile** [comment]
If we follow the compiler error, we can try to use `Box<dyn Bundle>`. However, this fails as `Bundle` is not object safe.

```rust
#[test]
fn impl_boxed_bundle_return_type() {
    fn spawn_bundle_naive(strategy: &ComponentStrategy) -> Box<dyn Bundle> {
        match strategy {
            ComponentStrategy::A => Box::new((A,)),
            ComponentStrategy::B => Box::new((B,)),
            ComponentStrategy::AAndB => Box::new((A, B)),
        }
    }
}
```

**Comment by alice-i-cecile** [comment]
We can brute force this, by operating on the world directly. This works, but requires blocking access.

```rust
#[test]
fn exclusive_world_access() {
    fn spawn_bundle_exclusive_world_access(
        world: &mut World,
        strategy: ComponentStrategy,
    ) -> Entity {
        let mut entity_world_mut = world.spawn_empty();

        match strategy {
            ComponentStrategy::A => entity_world_mut.insert(A).id(),
            ComponentStrategy::B => entity_world_mut.insert(B).id(),
            ComponentStrategy::AAndB => entity_world_mut.insert(A).insert(B).id(),
        }
    }

    let mut world = World::new();
    spawn_bundle_exclusive_world_access(&mut world, ComponentStrategy::A);
    spawn_bundle_exclusive_world_access(&mut world, ComponentStrategy::B);
    spawn_bundle_exclusive_world_access(&mut world, ComponentStrategy::AAndB);
}
```

**Comment by alice-i-cecile** [comment]
We could also pass in a mutable reference to `Commands`.

This can be used in an ordinary system, but requires plumbing through the `Commands` reference.
`Commands` references are also overly powerful: they can be used to modify the world in any way.

```rust
#[test]
fn ref_mut_commands() {
    use bevy::ecs::system::RunSystemOnce;

    fn spawn_bundle_ref_mut_commands(
        commands: &mut Commands,
        strategy: ComponentStrategy,
    ) -> Entity {
        let mut entity_commands = commands.spawn_empty();

        match strategy {
            ComponentStrategy::A => entity_commands.insert(A).id(),
            ComponentStrategy::B => entity_commands.insert(B).id(),
            ComponentStrategy::AAndB => entity_commands.insert(A).insert(B).id(),
        }
    }

    let mut world = World::new();

    fn my_system(mut commands: Commands) {
        spawn_bundle_ref_mut_commands(&mut commands, ComponentStrategy::A);
        spawn_bundle_ref_mut_commands(&mut commands, ComponentStrategy::B);
        spawn_bundle_ref_mut_commands(&mut commands, ComponentStrategy::AAndB);
    }

    world.run_system_once(my_system);
}
```

**Comment by alice-i-cecile** [comment]
We can restrict the power of the `Commands` reference by using `EntityCommands`.

This requires a bit more boilerplate, as you must spawn your entity first, but is more restrictive.
There are some gnarly liftimes involved, but we get an `EntityCommands` back out,
allowing us to chain commands together in the ordinary way.

```rust
#[test]
fn ref_mut_entity_commands() {
    use bevy::ecs::system::{EntityCommands, RunSystemOnce};

    fn spawn_bundle_ref_mut_entity_commands<'arg, 'w, 's, 'a>(
        commands: &'arg mut EntityCommands<'w, 's, 'a>,
        strategy: ComponentStrategy,
    ) -> &'arg mut EntityCommands<'w, 's, 'a> {
        match strategy {
            ComponentStrategy::A => commands.insert(A),
            ComponentStrategy::B => commands.insert(B),
            ComponentStrategy::AAndB => commands.insert(A).insert(B),
        }
    }

    let mut world = World::new();

    fn my_system(mut commands: Commands) {
        let mut entity_a = commands.spawn_empty();
        spawn_bundle_ref_mut_entity_commands(&mut entity_a, ComponentStrategy::A);

        let mut entity_b = commands.spawn_empty();
        spawn_bundle_ref_mut_entity_commands(&mut entity_b, ComponentStrategy::B);

        let mut entity_a_and_b = commands.spawn_empty();
        spawn_bundle_ref_mut_entity_commands(&mut entity_a_and_b, ComponentStrategy::AAndB);
    }

    world.run_system_once(my_system);
}
```

**Comment by alice-i-cecile** [comment]
We can clean up the `EntityCommands` pattern above, by using an extension method on `EntityCommands`.

```rust
#[test]
fn entity_commands_simple_extension() {
    use bevy::ecs::system::{EntityCommands, RunSystemOnce};

    trait EntityCommandsExt {
        fn spawn_bundle_by_strategy(&mut self, strategy: ComponentStrategy) -> &mut Self;
    }

    impl EntityCommandsExt for EntityCommands<'_, '_, '_> {
        fn spawn_bundle_by_strategy(&mut self, strategy: ComponentStrategy) -> &mut Self {
            match strategy {
                ComponentStrategy::A => self.insert(A),
                ComponentStrategy::B => self.insert(B),
                ComponentStrategy::AAndB => self.insert(A).insert(B),
            }
        }
    }

    let mut world = World::new();

    fn my_system(mut commands: Commands) {
        let mut entity_a = commands.spawn_empty();
        entity_a.spawn_bundle_by_strategy(ComponentStrategy::A);

        let mut entity_b = commands.spawn_empty();
        entity_b.spawn_bundle_by_strategy(ComponentStrategy::B);

        let mut entity_a_and_b = commands.spawn_empty();
        entity_a_and_b.spawn_bundle_by_strategy(ComponentStrategy::AAndB);
    }

    world.run_system_once(my_system);
}
```

**Comment by alice-i-cecile** [comment, marked as answer]
However, the direct `EntityCommands` extension trait isn't very flexible: now we need a different extension method for every place we want to use this pattern!

Instead, let's pass in a closure into our extension method, which controls which builder we're using.

This approach has an elaborate setup, but very flexible and quite comfortable to use. Critically, by setting your `Config` type to something more complex, we can pass in things like `ResMut<Assets<Mesh>>` or `Handle<ColorMaterial>`.

```rust
#[test]
fn entity_commands_closure_extension() {
    use bevy::ecs::system::{EntityCommands, RunSystemOnce};

    trait EntityCommandsExt<Config> {
        fn spawn_dynamic_bundle(
            &mut self,
            config: Config,
            f: impl FnOnce(Config, &mut Self),
        ) -> &mut Self;
    }

    impl<Config> EntityCommandsExt<Config> for EntityCommands<'_, '_, '_> {
        fn spawn_dynamic_bundle(
            &mut self,
            config: Config,
            f: impl FnOnce(Config, &mut Self),
        ) -> &mut Self {
            f(config, self);
            self
        }
    }

    fn my_dynamic_builder(strategy: ComponentStrategy, commands: &mut EntityCommands<'_, '_, '_>) {
        match strategy {
            ComponentStrategy::A => commands.insert(A),
            ComponentStrategy::B => commands.insert(B),
            ComponentStrategy::AAndB => commands.insert(A).insert(B),
        };
    }

    let mut world = World::new();

    fn my_system(mut commands: Commands) {
        let mut entity_a = commands.spawn_empty();
        entity_a.spawn_dynamic_bundle(ComponentStrategy::A, my_dynamic_builder);

        let mut entity_b = commands.spawn_empty();
        entity_b.spawn_dynamic_bundle(ComponentStrategy::B, my_dynamic_builder);

        let mut entity_a_and_b = commands.spawn_empty();
        entity_a_and_b.spawn_dynamic_bundle(ComponentStrategy::AAndB, my_dynamic_builder);
    }

    world.run_system_once(my_system);
}
```

**Comment by lucidBrot** [comment]
The initial post captures my problem well, but I'm failing to understand how the solution helps. Maybe I'm structuring something fundamentally wrong in my own code. Or maybe this pattern was not meant for use by bevy _users_, only internally?  
Could you explain how I am supposed to do this?:

I parse in some save information from a text file. The parse function should return a list of `Bundle` that I later spawn into the world. After all the things have been parsed by various functions, I `spawn` them. I do not do the spawning immediately, because I want to first be sure that all parsing steps succeeded.   
This works as long as every parse function returns only one type of Bundle. But depending on the text file contents, the Bundle shall sometimes contain `CheckpointIsUnlocked` as a Component (so I can make systems query for unlocked checkpoints).  

This idea does not work though, because like you say in the first post, `impl Bundle` is not a valid return type when there are different bundle types. And `dyn Bundle` does not work because `Bundle` is not a object-safe. So the initial problem I have is exactly what you describe.

But if I were to use your solution with `spawn_dynamic_bundle`, I would have to know implementation details of the Checkpoint at the line where I spawn the Bundle. This gains me nothing over calling two separate parse functions - one for unlocked checkpoints and one for locked checkpoints.
(Which falls apart quickly once there are more than one such field)  

So I have been thinking perhaps this can be done more cleanly with [BundleEffects](https://bevy.org/news/bevy-0-16/#:~:text=%22bundle%20effects%22%2C%20which%20are%20applied%20immediately%20after%20a%20Bundle%20is%20inserted.)  
> "bundle effects", which are applied immediately after a Bundle is inserted   

by storing this knowledge in the Bundle's Effect instead.  

but considering that [DynamicBundle](https://docs.rs/bevy/latest/bevy/ecs/bundle/trait.DynamicBundle.html) needs unsafe to be implemented, I suspect this is not what I am supposed to be doing.   


I hope I could make my use-case and confusion clear. If this scenario is something where your pattern is supposed to simplify usage, I'd appreciate a hint.
